NetCore/Resize.cs

```diff
@@ -141,7 +149,7 @@ public SKSize SkiaBitmapResizeBenchmark()
             const double xFactor = (double)ResizedWidth / Width;
             const double yFactor = (double)ResizedHeight / Height;
 
-            using (var original = NetVips.Image.Black(Width, Height))
+            using (var original = NetVips.Image.Black(Width, Height).CopyMemory())
```

Comment:
This was erroneously removed in https://github.com/bleroy/core-imaging-playground/pull/17.  Per https://github.com/bleroy/core-imaging-playground/pull/17#issuecomment-542381464 (emphasis mine):

> Re. the operation cache: all libvips operations are functional, that is, they have no side-effects. This means **you can replace the result of any operation with the result of a previous call with the same arguments**. By default, libvips aims to keep about 100mb of cached operations.

> This can give misleading benchmark results if you loop several times over the same image. **I usually try to have N distinct input images rather than looping N times over the same image**.

I'm restoring it here to ensure Vips is not able to see that `Image.Black` is re-used between iterations.  @kleisauke is there a cheaper way to accomplish the same?

Reply:
You can try to disable libvips' operation cache instead (as mentioned in https://github.com/bleroy/core-imaging-playground/pull/17#issuecomment-542376960). For example:
<details>
 <summary>Details</summary>

```diff
diff --git a/NetCore/Resize.cs b/NetCore/Resize.cs
index 1111111..2222222 100644
--- a/NetCore/Resize.cs
+++ b/NetCore/Resize.cs
@@ -6,6 +6,7 @@ using BenchmarkDotNet.Attributes;
 using FreeImageAPI;
 using ImageMagick;
 using NetVips;
+using NetVipsUtil = NetVips.NetVips;
 using PhotoSauce.MagicScaler;
 using SixLabors.ImageSharp;
 using SixLabors.ImageSharp.Processing;
@@ -24,7 +25,13 @@ namespace ImageProcessing
         private const int ResizedWidth = 150;
         private const int ResizedHeight = 99;
 
-        public Resize() => OpenCL.IsEnabled = false;
+        public Resize()
+        {
+            OpenCL.IsEnabled = false;
+
+            // Disable libvips' operation cache to ensure tests are as fair as they can be
+            NetVipsUtil.CacheSetMax(0);
+        }
 
         [Benchmark(Baseline = true, Description = "System.Drawing Resize")]
         public Size ResizeSystemDrawing()
@@ -146,7 +153,7 @@ namespace ImageProcessing
             {
                 // libvips is "lazy" and will not process pixels
                 // until you write to an output file, buffer or memory
-                var _ = resized.WriteToMemory();
+                var _ = resized.CopyMemory();
 
                 return (resized.Width, resized.Height);
             }
```
</details>


I also changed `WriteToMemory()` back to `CopyMemory()` within this diff to [avoid having to copy the whole image to a managed array](https://github.com/kleisauke/net-vips/blob/adb5860d400a1a7f6397cb6f2b148d037335299e/src/NetVips/Image.cs#L1072-L1073), sorry that I didn't catch this earlier.

Reply:
Yep, the `WriteToMemory`->`CopyMemory` change drops the managed memory allocation from 18.7KB to 3.7KB, which makes perfect sense.

Removing the `CopyMemory` from `Image.Black`, with or without `CacheSetMax(0)` results in the benchmark time dropping from 9.6ms to 4.7ms (on the RPi), which doesn't make sense to me at all.  Unless I'm completely misunderstanding how it works, all that `CopyMemory` should be doing is allocating 3MB of native memory and then zeroing it, which can't possibly take longer than running a bicubic resample over that same memory -- which in turn means it can't make the benchmark time more than double.  I suspect regardless of the cache setting, there's something being re-used between iterations unless the image memory is made opaque to Vips, but I don't know enough about how it works to be sure 🤷.

Reply:
I've just been reading through the Vips docs, and I see now that because `Image.Black()` is called with only a width and height, it's creating a single-channel (8 bits per pixel) image.  That should be `Image.Black(Width, Height, 3)` to create an RGB image, which nearly doubles the timings (8.7ms without `CopyMemory`, 16.1ms with).

Reply:
Ah yes, that makes sense. It will indeed have to create a 3-channel image here. 8.7ms sound reasonable for a RPi 4.

I think the problem with `CopyMemory` is that for partial images (i.e. `VIPS_IMAGE_PARTIAL`) the whole pipeline is executed, see:
https://github.com/libvips/libvips/blob/a43e9d3982cd371479e6ac5ae9385723c7ef81bb/libvips/iofuncs/image.c#L3345-L3351

That could explain why the timing is doubling.

/cc @jcupitt

Reply:
But isn't the whole pipeline at that point just `Image.Black`?  I was assuming `CopyMemory` caused it to start a new pipeline, which I figured would break the caching since the materialized black image then becomes opaque memory which could have been modified outside libvips.

Reply:
Hello everyone, without the first `CopyMemory()` libvips will cache and reuse the whole pipeline, so it'll only execute the resize once and then just give you the same result N times. The `CopyMemory` creates a a unique pointer at the start of the pipeline for each call, so it'll mean the resize is called differently, so therefore the resize will execute N times.

You can see this happening with a tiny benchmark (python for convenience):

```python
#!/usr/bin/python3

import pyvips
import logging

logging.basicConfig(level=logging.INFO)

for i in range(100):
    x = pyvips.Image.black(1000, 1000)
    y = x.resize(0.1)
    z = y.copy_memory()
```

If you run that, you'll see:

```
john@kiwi:~/try$ ./copymem.py
INFO:pyvips:VIPS: shrinkv by 5
INFO:pyvips:VIPS: shrinkh by 5
INFO:pyvips:VIPS: residual reducev by 0.5
INFO:pyvips:VIPS: reducev: 13 point mask
INFO:pyvips:VIPS: reducev: using vector path
INFO:pyvips:VIPS: residual reduceh by 0.5
INFO:pyvips:VIPS: reduceh: 13 point mask
john@kiwi:~/try$ 
```

So although that looped 100 times, the resize only executed once. Add a `copy_memory` at the top and you'll see a *lot* of output, and runtime will roughly double.